File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.py
from collections import deque


def spinning_90(board):
    result = [[0] * len(board) for _ in range(len(board[0]))]
    for r in range(len(board[0])):
        for c in range(len(board)):
            result[r][c] = board[c][len(board[0]) - 1 - r]

    return result


def spinning_180(board):
    return spinning_90(spinning_90(board))


def spinning_270(board):
    return spinning_180(spinning_90(board))


def is_equals(cnt_a, table_a, cnt_b, table_b):
    if cnt_a < cnt_b:
        return False

    r_a = len(table_a)
    r_b = len(table_b)

    c_a = len(table_a[0])
    c_b = len(table_b[0])

    if r_a != r_b or c_a != c_b:
        return False

    for r in range(r_a):
        for c in range(c_a):
            if table_b[r][c] != table_a[r][c]:
                return False

    return True


def equals(cnt_a, table_a, cnt_b, table_b):
    spinning1 = spinning_90(table_b)
    spinning2 = spinning_90(spinning1)
    spinning3 = spinning_90(spinning2)

    for s in [table_b, spinning1, spinning2, spinning3]:
        if is_equals(cnt_a, table_a, cnt_b, s):
            return True

    return False


def bfs(y, x, visited, game_board, target):
    dy = [0, -1, 0, 1]
    dx = [-1, 0, 1, 0]

    puzzle = []

    min_y, min_x, max_y, max_x = y, x, y, x

    que = deque()
    que.append((y, x))
    visited[y][x] = True
    cnt = 0
    while que:
        y, x = que.popleft()
        cnt += 1
        puzzle.append((y, x))
        min_y = min(min_y, y)
        max_y = max(max_y, y)
        min_x = min(min_x, x)
        max_x = max(max_x, x)

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if not (0 <= ny < len(visited) and 0 <= nx < len(visited[0])):
                continue

            if visited[ny][nx] or game_board[ny][nx] != target:
                continue

            visited[ny][nx] = True
            que.append((ny, nx))
    result = [[0] * (max_x - min_x + 1) for _ in range(max_y - min_y + 1)]
    for y, x in puzzle:
        result[y - min_y][x - min_x] = 1

    return cnt, result


def solution(game_board, table):
    # game_board = [[1, 1, 0], [0, 1, 1]]
    visited_a = [[False] * len(game_board[0]) for _ in range(len(game_board))]
    visited_b = [[False] * len(game_board[0]) for _ in range(len(game_board))]
    blank = []
    puzzle = []
    for r in range(len(game_board)):
        for c in range(len(game_board[0])):
            if not visited_a[r][c] and game_board[r][c] == 0:
                blank.append(bfs(r, c, visited_a, game_board, 0))

            if not visited_b[r][c] and table[r][c] == 1:
                puzzle.append(bfs(r, c, visited_b, table, 1))

    visited_b = set()
    visited_p = set()
    cnt = 0
    for i in range(len(blank)):
        for j in range(len(puzzle)):
            b = blank[i]
            p = puzzle[j]
            if i not in visited_b and j not in visited_p and equals(*b, *p):
                visited_b.add(i)
                visited_p.add(j)
                cnt += b[0]

    return cnt

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;

class Table {
    public int cnt;
    public int row;
    public int column;
    public int[][] data;

    public Table spinning() {
        int[][] newData = new int[column][row];
        for (int r = 0; r < column; r++) {
            for (int c = 0; c < row; c++) {
                newData[r][c] = data[c][column - 1 - r];
            }
        }

        return new Table(newData, column, row, cnt);
    }

    public boolean equals(Table other) {
        Table spin1 = other.spinning();
        Table spin2 = spin1.spinning();
        Table spin3 = spin2.spinning();

        return (this.compare(other) || this.compare(spin1) || this.compare(spin2) || this.compare(spin3));
    }

    public boolean compare(Table other) {
        if (row != other.row || column != other.column) {
            return false;
        }

        for (int r = 0; r < row; r++) {
            for (int c = 0; c < column; c++) {
                if (data[r][c] != other.data[r][c]) {
                    return false;
                }
            }
        }

        return true;
    }

    @Override
    public String toString() {
        return "Table{" +
                "cnt=" + cnt +
                ", row=" + row +
                ", column=" + column +
                ", data=" + Arrays.toString(data) +
                '}';
    }

    public Table(int[][] data, int row, int column, int cnt) {
        this.data = data;
        this.row = row;
        this.column = column;
        this.cnt = cnt;
    }

    public Table(int y, int x, int[][] visited, int[][] table, int target) {
        LinkedList<int[]> queue = new LinkedList<>();
        ArrayList<int[]> puzzle = new ArrayList<>();
        queue.addLast(new int[]{y, x});
        visited[y][x] = 1;
        int[] dy = new int[]{0, -1, 0, 1};
        int[] dx = new int[]{-1, 0, 1, 0};

        int minY = y;
        int maxY = y;
        int minX = x;
        int maxX = x;

        while (!queue.isEmpty()) {
            int[] ints = queue.pollFirst();
            cnt += 1;
            y = ints[0];
            x = ints[1];

            puzzle.add(ints);

            minY = Math.min(minY, y);
            maxY = Math.max(maxY, y);
            minX = Math.min(minX, x);
            maxX = Math.max(maxX, x);

            for (int i = 0; i < 4; i++) {
                int ny = y + dy[i];
                int nx = x + dx[i];


                if (!(0 <= ny && ny < table.length && 0 <= nx && nx < table[0].length)) continue;

                if (visited[ny][nx] == 1 || table[ny][nx] != target) continue;

                visited[ny][nx] = 1;
                queue.addLast(new int[]{ny, nx});
            }
        }

        row = maxY - minY + 1;
        column = maxX - minX + 1;
        data = new int[row][column];

        for (int[] ints : puzzle) {
            int r = ints[0];
            int c = ints[1];

            data[r - minY][c - minX] = 1;
        }
    }
}

class Solution {
    public int solution(int[][] game_board, int[][] table) {
        int[][] visitedA = new int[game_board.length][game_board[0].length];
        int[][] visitedB = new int[table.length][table[0].length];

        ArrayList<Table> blank = new ArrayList<>();
        ArrayList<Table> puzzle = new ArrayList<>();

        for (int r = 0; r < game_board.length; r++) {
            for (int c = 0; c < game_board[0].length; c++) {
                if (visitedA[r][c] == 0 && game_board[r][c] == 0) {
                    blank.add(new Table(r, c, visitedA, game_board, 0));
                }

                if (visitedB[r][c] == 0 && table[r][c] == 1) {
                    puzzle.add(new Table(r, c, visitedB, table, 1));
                }
            }
        }

        Table[] blankArray = blank.toArray(Table[]::new);
        Table[] puzzleArray = puzzle.toArray(Table[]::new);


        HashSet<Integer> setA = new HashSet<>();
        HashSet<Integer> setB = new HashSet<>();

        int cnt = 0;
        for (int i = 0; i < blankArray.length; i++) {
            for (int j = 0; j < puzzleArray.length; j++) {
                Table b = blankArray[i];
                Table p = puzzleArray[j];

                if (!setA.contains(i) && !setB.contains(j) && b.equals(p)) {
                    setA.add(i);
                    setB.add(j);
                    cnt += b.cnt;
                }
            }
        }
        return cnt;
    }
}